_posts/2022-10-01-arrow-parquet-encoding-part-1.md

```diff
@@ -0,0 +1,152 @@
+---
+layout: post
+title: Arrow and Parquet Part 1: Primitive Types and Nullability
+date: "2022-10-01 00:00:00"
+author: tustvold, alamb
+categories: [parquet, arrow]
+---
+<!--
+{% comment %}
+Licensed to the Apache Software Foundation (ASF) under one or more
+contributor license agreements.  See the NOTICE file distributed with
+this work for additional information regarding copyright ownership.
+The ASF licenses this file to you under the Apache License, Version 2.0
+(the "License"); you may not use this file except in compliance with
+the License.  You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+{% endcomment %}
+-->
+
+## Introduction
+
+We recently completed a long-running project within [Rust Apache Arrow](https://github.com/apache/arrow-rs) to complete support for reading and writing arbitrarily nested Parquet and Arrow schemas. This is a complex topic, and we encountered a lack of approachable technical information, and thus wrote this blog to share our learnings with the community.
+
+[Apache Arrow](https://arrow.apache.org/) is an open, language-independent columnar memory format for flat and hierarchical data, organized for efficient analytic operations. [Apache Parquet](https://parquet.apache.org/) is an open, column-oriented data file format designed for very efficient data encoding and retrieval.
+
+It is increasingly common for analytic systems to use Arrow to process data stored in Parquet files, and therefore fast, efficient, and correct translation between them is a key building block.
+
+Historically analytic processing primarily focused on querying data with a tabular schema, where there are a fixed number of columns, and each row contains a single value for a each column. However, with the increasing adoption of structured document formats such as XML, JSON, etc…, this schema limitation can seem antiquated and unnecessarily limiting.
```

Comment:
```suggestion
Historically analytic processing primarily focused on querying data with a tabular schema, where there are a fixed number of columns, and each row contains a single value for each column. However, with the increasing adoption of structured document formats such as XML, JSON, etc…, this schema limitation can seem antiquated and unnecessarily limiting.
```

Reply:
(just removing the extra `a`)

Reply:
but, now that I read the next paragraph, it feels like a cliffhanger to say that the schema limitation "can seem antiquated and [limiting]." and not address whether it is/isn't antiquated/limiting or where you might discuss such a topic.

If that's not covered in the blog post, it seems better to just say that structured document formats are an alternative that do not expect a schema describing the structure.

Reply:
but I don't think using XML/JSON requires flattening of a source schema. Or, maybe you're thinking in the direction of going from JSON/XML -> tabular?

I think I would prefer them to both be introduced neutrally. Historically tabular schemas have been used, but JSON and XML are 2 popular alternate formats that do not need a schema.

Reply:
> Or, maybe you're thinking in the direction of going from JSON/XML -> tabular?

Yes that is precisely the transformation that has been a repeated pain point in my experience of data pipelines

Reply:
What do you think of

```suggestion
Historically analytic processing primarily focused on querying data with a tabular schema, where there are a fixed number of columns, and each row contains a single value for each column. However, with the increasing adoption of structured document formats such as XML, JSON, etc…, this schema limitation can be frustrating for users, as it necessitates often non-trivial data transformation to first flatten the document data into a tabular format.
```

Reply:
That seems reasonable.

IMO, the use of "Historically" and "schema limitation" makes it seem like you're trying to suggest that schemas can be frustrating. Maybe this, or some things from it, could be an improvement:

```suggestion
Historically, analytic processing primarily focused on querying data with a tabular schema. Recently, there has been an increasing adoption of structured document formats such as XML or JSON. For structured document formats, specifying a schema can seem limiting. For tabular data, non-trivial data transformations to flatten document data is often required. For these reasons, translating between the two can be frustrating for users.
```

Reply:
I think that is more confusing as it seems to suggest that document formats don't have schema??

Reply:
Thanks for sticking with this -- nice improvement